File: webapp/frontend/src/api/convert.ts
import type { ConvertRequest, ConvertResult } from "../types/convert";

const BASE_URL = "";

export async function postConvert(request: ConvertRequest): Promise<ConvertResult> {
    const form = new FormData();
    form.append("file", request.file);
    form.append("catalog", request.catalog);
    form.append("schema", request.schema);
    if (request.source_catalog) form.append("source_catalog", request.source_catalog);
    if (request.source_schema) form.append("source_schema", request.source_schema);
    if (request.prefix) form.append("prefix", request.prefix);
    if (request.fact_tables) form.append("fact_tables", request.fact_tables);
    if (request.exclude_tables) form.append("exclude_tables", request.exclude_tables);
    form.append("include_isolated", String(request.include_isolated ?? false));
    form.append("dialect", request.dialect ?? "databricks");

    const res = await fetch(`${BASE_URL}/api/convert`, {
        method: "POST",
        body: form,
    });

    if (!res.ok) {
        let detail = res.statusText;
        try {
            const body = await res.json();
            detail = body.detail ?? detail;
        } catch {
            // ignore JSON parse failure
        }
        throw new Error(`${res.status}: ${detail}`);
    }

    return res.json() as Promise<ConvertResult>;
}

File: webapp/backend/app/routes/convert.py
"""Route for uploading and converting a PowerBI PBIX file."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Form, UploadFile

from ..models import ConvertOut
from ..services import converter

router = APIRouter(prefix="/api/convert", tags=["convert"])


@router.post("", response_model=ConvertOut)
async def convert_pbix(
    file: UploadFile,
    catalog: str = Form(...),
    schema: str = Form(...),
    source_catalog: Optional[str] = Form(default=None),
    source_schema: Optional[str] = Form(default=None),
    prefix: str = Form(default=""),
    fact_tables: Optional[str] = Form(default=None),
    exclude_tables: Optional[str] = Form(default=None),
    include_isolated: bool = Form(default=False),
    dialect: str = Form(default="databricks"),
) -> ConvertOut:
    """Upload a .pbix file and convert it to Databricks metric views.

    Returns structured output describing each generated metric view, the DAX → SQL
    translations performed, M-expression resolutions, and any warnings.
    """
    pbix_bytes = await file.read()
    filename = file.filename or "upload.pbix"

    return converter.run_convert(
        pbix_bytes=pbix_bytes,
        pbix_filename=filename,
        catalog=catalog,
        schema=schema,
        source_catalog=source_catalog or None,
        source_schema=source_schema or None,
        fact_tables=fact_tables or None,
        prefix=prefix,
        exclude_tables=exclude_tables or None,
        include_isolated=include_isolated,
        dialect=dialect,
    )
